Бравилов Николай

kol9nus opened the pull request.

kol9nus committed:
Complete quest

honest-hrundel commented:
:tomato: Пройдено тестов 23 из 24

kol9nus committed:
Add length check

kol9nus committed:
Refactor code

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

kol9nus committed:
Refactor code a bit more

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

kol9nus commented:
🍏

Mokoshka commented on lib.js:
```diff
@@ -19,34 +118,48 @@ function Iterator(friends, filter) {
  */
 function Filter() {
-    console.info('Filter');
+    this.genderPattern = /[\s\S]*/;
```
Разве `[\s\S]*` не эквивалентно `.*`?

kol9nus replied:
Туда не входит перевод строки. Маловероятно конечно, но по условию задания написано

> По умолчанию такой фильтр никого не отсеивает

Т.е. соответствует вообще любой строке.

Mokoshka replied:
Хм... Я сделала что-то не так?
![image](https://cloud.githubusercontent.com/assets/14892550/20607641/ef28172c-b29b-11e6-9ea1-9968e05ccf79.png)

kol9nus replied:
Это у меня (идейно) неточные регвыры были. Так наглядней 
![image](https://cloud.githubusercontent.com/assets/11522574/20607930/67b32964-b29e-11e6-9551-8b9cc894a963.png)

kol9nus replied:
Но в данном случае да, `.*` будет выглядеть куда лучше.

Mokoshka commented on lib.js:
```diff
@@ -7,9 +7,108 @@
+    this.levelIndex++;
+    while (
+    this.levelIndex < this.friendshipLevels.length &&
+    this.friendshipLevels[this.levelIndex].length === 0
```
Ой-ой. Как минимум тут не хватает отступов и лучше написать
```
while(condition1 &&
    condition2) { ... }
```

Mokoshka commented on lib.js:
```diff
@@ -7,9 +7,108 @@
+
+    var getFriendByName = createFriendByNameGetter();
+    this.friendshipLevels = divideIntoFriendshipLevels().map(function (level) {
+        return filter.filter(level)
```
```
filter
    .filter()
    .sort()
```
Так лучше будет

Mokoshka commented on lib.js:
```diff
@@ -7,9 +7,108 @@
+        });
+
+        /* Запоминаем, что уже брали их */
+        var viewedFriends = Object.create(null);
```
О_О
А почему не просто `{}`?

kol9nus replied:
В [лекциях](https://github.com/urfu-2015/javascript-lectures/blob/master/8-constructors.md#objectcreatenull-) было сказано:

> Такого рода объекты, часто называют словарями, так как их удобно использовать для хранения данных, они лишены особенностей, связанных с делегированием свойств по цепочке прототипов. По ним можно безопасно итерироваться циклом for..in.

Вот и решил, что тут это определение полностью подходит.

Mokoshka replied:
Окей

Mokoshka commented on lib.js:
```diff
@@ -7,9 +7,108 @@
+
+        /* Запоминаем, что уже брали их */
+        var viewedFriends = Object.create(null);
+        queueForAdding.forEach(function (friend) {
```
Тут можно на reduce переписать и передать туда пустой объект

kol9nus replied:
А почему reduce предпочтительнее? Через него было, но forEach выглядит как-то "элегантнее". Для сравнения:

```
var viewedFriends = Object.create(null);
queueForAdding.forEach(function (friend) {
    viewedFriends[friend.name] = true;
});
```
И через reduce:

```
var viewedFriends = queueForAdding.reduce(function (viewed, friend) {
    viewed[friend.name] = true;
            
    return viewed;
}, Object.create(null));
```

Mokoshka replied:
Использование reduce по смыслу больше подходит. Читаешь код, видишь reduce - ага, значит здесь что-то накапливается по такому вот правилу

Mokoshka commented on lib.js:
```diff
@@ -7,9 +7,108 @@
+};
+
+Iterator.prototype.next = function () {
+    return !this.done() ? this.friendshipLevels[this.levelIndex][this.nextIndex++] : null;
```
`this.done() ? null : this.friendshipLevels[this.levelIndex][this.nextIndex++]`?

kol9nus committed:
Fix remarks

filter.filter was renamed to filter.filterFriends to differ that
function from Array.prototype.filter

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

mokhov commented on lib.js:
```diff
@@ -19,34 +117,48 @@ function Iterator(friends, filter) {
  */
 function FemaleFilter() {
-    console.info('FemaleFilter');
+    this.genderPattern = /^female$/;
```
А просто сравнить со строкой будет не проще чем писать регулярки? :)

kol9nus replied:
Проще, просто не знал, как сделать сравнение для Filter так, чтобы независимо от значения поля gender с данной функцией фильтрации (filterFriends) (другую фильтрацию как-то не придумал) всё проходило, а тут просто переменную поменять, не трогая функцию. На ум пришли только регулярки.

kol9nus committed:
Fix remarks

honest-hrundel commented:
:green_apple: Пройдено тестов 24 из 24

kol9nus commented:
🍏

Mokoshka commented:
:rocket:

mokhov commented on lib.js:
```diff
@@ -7,9 +7,109 @@
+        }, Object.create(null));
+
+        var numberOfFriendsInNextLevel;
+        while ((numberOfFriendsInNextLevel = queueForAdding.length)) {
```
Лучше сделать этот код читаемее, т.е сделать переменную-флаг, чем писать `((expression))`

mokhov commented on lib.js:
```diff
@@ -7,9 +7,109 @@
+            dict[friend.name] = friend;
+
+            return dict;
+        }, Object.create(null));
```
Почему здесь не просто `{}`?

kol9nus replied:
В [лекциях](https://github.com/urfu-2015/javascript-lectures/blob/master/8-constructors.md#objectcreatenull-) было сказано:

> Такого рода объекты, часто называют словарями, так как их удобно использовать для хранения данных, они лишены особенностей, связанных с делегированием свойств по цепочке прототипов. По ним можно безопасно итерироваться циклом for..in.

Вот и решил, что тут это определение полностью подходит.
